### ISSUE TITLE
Not compiling with gcc >= 9

### ISSUE THREAD

**Issue by danhper** [ethereum/aleth#5637]
Hi,

gcc 9 seems to introduce warnings which make the build fail.
In particular, `-Wdeprecated-copy`, which forbids to use `Foo operator=` is `Foo& operator=` is defined generates a couple of errors in `libdevcore/FixedHash.h` because

```cpp
SecureFixedHash<T>& operator=(SecureFixedHash<T> const& _c)
```

is defined but `SecureFixedHash<T> operator=` is used here and there.

Another warning newly triggered is `stringop-truncation` because of `strncpy` at in `libweb3jsonrpc/UnixSocketServer.cpp:UnixDomainSocketServer::StartListening`.

I am not sure what is the best way to go so for now, I simply disabled the warnings, globally for `deprecated-copy`:

```cmake
# cmake/EthCompilerSettings.cmake
add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-deprecated-copy>)
```

and locally for `stringop-truncation`:

```cpp
// libweb3jsonrpc/UnixSocketServer.cpp

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wstringop-truncation"
strncpy(m_address.sun_path, m_path.c_str(), c_socketPathMaxLength);
#pragma GCC diagnostic pop
```

More information about the new warnings:
https://gcc.gnu.org/gcc-9/changes.html

**Comment by moneroexamples** [comment]
And GTEST builds for you? I can't even get to this point you are describing when default because gtest build fails.

have to build without tests to get to your issue:

```
cmake -DTESTS=OFF ..
```

**Comment by moneroexamples** [comment]
Once you pass the `FixedHash<T>` error, there is second one:

```
/home/mwo2/aleth/libweb3jsonrpc/UnixSocketServer.cpp: In member function ‘virtual bool dev::UnixDomainSocketServer::StartListening()’:
/home/mwo2/aleth/libweb3jsonrpc/UnixSocketServer.cpp:85:10: error: ‘char* strncpy(char*, const char*, size_t)’ specified bound 108 equals destination size [-Werror=stringop-truncation]
   85 |   strncpy(m_address.sun_path, m_path.c_str(), c_socketPathMaxLength);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [libweb3jsonrpc/CMakeFiles/web3jsonrpc.dir/build.make:219: libweb3jsonrpc/CMakeFiles/web3jsonrpc.dir/UnixSocketServer.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:989: libweb3jsonrpc/CMakeFiles/web3jsonrpc.dir/all] Error 2
```

**Comment by danhper** [comment]
Good point, I forgot to mention that I also had to disable the tests for now.
For the error you mention, the temporary fix is in the issue. You need to disable `stringop-truncation` locally in `libweb3jsonrpc/UnixSocketServer.cpp` (or globally if you prefer).

**Comment by moneroexamples** [comment]
I patched the two issues  (not gtest problem):

```diff
diff --git a/libdevcore/FixedHash.h b/libdevcore/FixedHash.h
index 4a50e6ea8..b74129c4f 100644
--- a/libdevcore/FixedHash.h
+++ b/libdevcore/FixedHash.h
@@ -245,12 +245,16 @@ public:
 	explicit SecureFixedHash(byte const* _d, ConstructFromPointerType _t): FixedHash<T>(_d, _t) {}
 	~SecureFixedHash() { ref().cleanse(); }
 
-	SecureFixedHash<T>& operator=(SecureFixedHash<T> const& _c)
+    SecureFixedHash(SecureFixedHash<T> const& _c): FixedHash<T>(_c)
+    {
+		ref().cleanse();
+    }
+
+	SecureFixedHash<T>& operator=(SecureFixedHash<T> _c)
 	{
 		if (&_c == this)
 			return *this;
-		ref().cleanse();
-		FixedHash<T>::operator=(static_cast<FixedHash<T> const&>(_c));
+        std::swap(*this, _c);
 		return *this;
 	}
 
diff --git a/libweb3jsonrpc/UnixSocketServer.cpp b/libweb3jsonrpc/UnixSocketServer.cpp
index 9510f1a6c..5d1a3a2cb 100644
--- a/libweb3jsonrpc/UnixSocketServer.cpp
+++ b/libweb3jsonrpc/UnixSocketServer.cpp
@@ -82,7 +82,9 @@ bool UnixDomainSocketServer::StartListening()
 #ifdef __APPLE__
 		m_address.sun_len = m_path.size() + 1;
 #endif
-		strncpy(m_address.sun_path, m_path.c_str(), c_socketPathMaxLength);
+		strncpy(m_address.sun_path, m_path.c_str() - 1, c_socketPathMaxLength);
+        memset(m_address.sun_path, '\0', sizeof(m_address.sun_path));
+
 		::bind(m_socket, reinterpret_cast<sockaddr*>(&m_address), sizeof(sockaddr_un));
 		fs::permissions(m_path, fs::owner_read | fs::owner_write);
 		listen(m_socket, 128);
```

It compiles with this, though I'm not confident if what I did is the correct way of doing this. Would need to look more into what the fuctions/classes do.

**Comment by danhper** [comment]
For gtest, it compiled fine if I used my local installation instead of the one provided by hunter so I suspect Hunter is using an old version of gtest.

**Comment by etam** [comment]
I got this a bit differently:

```diff
Index: ethereum-cpp-1.6.0/libdevcore/FixedHash.h
===================================================================
--- ethereum-cpp-1.6.0.orig/libdevcore/FixedHash.h
+++ ethereum-cpp-1.6.0/libdevcore/FixedHash.h
@@ -244,6 +244,8 @@ public:
        explicit SecureFixedHash(std::string const& _s, ConstructFromStringType _t = FixedHash<T>::FromHex, ConstructFromHashType _ht = FixedHash<T>::FailIfDifferent): FixedHash<T>(_s, _t, _ht) {}
        explicit SecureFixedHash(byte const* _d, ConstructFromPointerType _t): FixedHash<T>(_d, _t) {}
        ~SecureFixedHash() { ref().cleanse(); }
+
+       SecureFixedHash(SecureFixedHash<T> const& _c): FixedHash<T>(_c) {}
 
        SecureFixedHash<T>& operator=(SecureFixedHash<T> const& _c)
        {
Index: ethereum-cpp-1.6.0/libweb3jsonrpc/UnixSocketServer.cpp
===================================================================
--- ethereum-cpp-1.6.0.orig/libweb3jsonrpc/UnixSocketServer.cpp
+++ ethereum-cpp-1.6.0/libweb3jsonrpc/UnixSocketServer.cpp
@@ -22,6 +22,7 @@ along with cpp-ethereum.  If not, see <h
 #if !defined(_WIN32)
 
 #include "UnixSocketServer.h"
+#include <algorithm>
 #include <sys/socket.h>
 #include <fcntl.h>
 #include <unistd.h>
@@ -57,7 +58,7 @@ fs::path getIpcPathOrDataDir()
 }
 
 UnixDomainSocketServer::UnixDomainSocketServer(string const& _appId):
-       IpcServerBase((getIpcPathOrDataDir() / fs::path(_appId + ".ipc")).string().substr(0, c_socketPathMaxLength))
+       IpcServerBase((getIpcPathOrDataDir() / fs::path(_appId + ".ipc")).string().substr(0, c_socketPathMaxLength - 1))
 {
 }
 
@@ -82,7 +83,7 @@ bool UnixDomainSocketServer::StartListen
 #ifdef __APPLE__
                m_address.sun_len = m_path.size() + 1;
 #endif
-               strncpy(m_address.sun_path, m_path.c_str(), c_socketPathMaxLength);
+               std::copy(m_path.begin(), m_path.end(), m_address.sun_path);
                ::bind(m_socket, reinterpret_cast<sockaddr*>(&m_address), sizeof(sockaddr_un));
                fs::permissions(m_path, fs::owner_read | fs::owner_write);
                listen(m_socket, 128);
```

**Comment by moneroexamples** [comment]
Just wondering if there is any progress on Hunter fixing x its googletest package as well as on the above errors as build is still broken on new systems, such as Arch linux.

**Comment by chfast** [comment]
Partly fixed by #5794.
Other fixes in #5798.
